### PR TITLE
Changelog 3.0

### PR DESCRIPTION
**Pull request by aartaka**
# Discussion

bd4c55b108c5bdaeee071437c104ed35da2b7683 removes pre-release sections. Should we do that?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] Git hygiene:
  - I have pulled from master before submitting this PR
  - There are no merge conflicts.
- [X] I've added the new dependencies as:
  - ASDF dependencies,
  - Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [x] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - I have updated the existing documentation to match my changes.
  - I have commented my code in hard-to-understand areas.
  - I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [X] Compilation and tests:
  - My changes generate no new warnings.
  - I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - I ran the tests locally (`(asdf:test-system :nyxt)` and `(asdf:test-system :nyxt/gi-gtk)`) and they pass.

### CODE REVIEW

**Comment by Ambrevar** [comment]
> #
> 
> [bd4c55b](https://github.com/atlas-engineer/nyxt/commit/bd4c55b108c5bdaeee071437c104ed35da2b7683) removes pre-release sections. Should we do that?

In my opinion we should: The changelog for pre-releases is only interesting to people using pre-releases.

**Comment by aartaka** [comment]
OK, merging then?

File: source/changelog.lisp [hunk 1]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+     (:li "When started with " (:code "--remote") " and without " (:code "--quit")
+          ", Nyxt now reads s-expression from standard input and sends it to the remote
+process. This avoids the performance penalty of a new process startup on each
+iteration.")
+     (:li (:nxref :function 'define-configuration)
+          " automatically resolves class names and slot names even without package prefix.")
+     (:li "Keybindings are also resolved to reasonable commands when necessary.")
+     (:li (:nxref :slot 'after-init-hook :class-name 'browser) " and "
+          (:nxref :slot 'after-startup-hook :class-name 'browser)
+          " are browser slots, instead of global variables they used to be.")
```
Comment (Ambrevar) [inline source/changelog.lisp:386]: Typo: "of THE global variables"

File: source/changelog.lisp [hunk 2]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
-(define-version "3-pre-release-5"
-  (:ul
-   (:li "Code listings on manual and help system are more readable.")
-   (:li "The REPL features " (:nxref :class-name 'nyxt/mode/repl:shell-cell "shell cells")
-        ", which pass commands to the underlying system shell.")
-   (:li "The prompt buffer tab interface has been updated. Tabs are stably sorted and do
-not shuffle around. It is now possible to middle click to delete tabs.")
-   (:li "Many prompts are filtered more aggressively, resulting in more exact suggestions."))
+     (:li "Search engines are now listed with their full name (when available).")
+     (:li "Internal buffers/pages are now rendered in the status area as: " (:code "internal") ".")))
```
Comment (Ambrevar) [inline source/changelog.lisp:398]: Spurious ":" after "as"

File: source/changelog.lisp [hunk 3]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+     (:ul
+      (:li (:nxref :macro 'define-class) " macro — to " (:code "nclasses") ".")
+      (:li "Portable GUI-friendly debugger — to " (:code "ndebug") ".")
+      (:li "JSON parsing — to " (:code "njson") ".")
+      (:li "Symbol search and listing — to " (:code "nsymbols") "."))
+     (:li "Move " (:code "lisp-system") " to separate repository as extension. The functions "
+          (:code "list-system") " and " (:code "load-system") " are no longer available.")
+     (:li "Move " (:code "os-package-manager-mode") " to separate repository as
+extension.")
+     (:li (:code "diff-mode") " is removed.")))
```
Comment (Ambrevar) [inline source/changelog.lisp:414]: Maybe "has been removed"

File: source/changelog.lisp [hunk 4]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
-has been revised and extended."))
-  (:h3 "Bug fixes")
-  (:ul
-   (:li "Update Flatpak build version.")
-   (:li "Extensions directory is more carefully checked to avoid crashes.")))
+  (:nsection
+    :title "Moves and removals"
+    (:ul
+     (:li "Move the " (:code "prompter") " library to a separate repository.")
+     (:li "Some of the Nyxt-internal logic was abstracted into separate libraries:")
```
Comment (Ambrevar) [inline source/changelog.lisp:404]: I would emphasize that for instance `(:code "nclasses")` refers to a Lisp system.

File: source/changelog.lisp [hunk 5]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+          (:code "list-system") " and " (:code "load-system") " are no longer available.")
+     (:li "Move " (:code "os-package-manager-mode") " to separate repository as
+extension.")
+     (:li (:code "diff-mode") " is removed.")))
+
+  (:nsection
+    :title "Prompt buffer improvements"
+    (:ul
+     (:li "New prompt buffer fuzzy matching algorithm, hopefully offering more relevant
+  results (Thanks to @BlueFlo0d!)")
```
Comment (Ambrevar) [inline source/changelog.lisp:420]: Missing trailing `.`.

File: source/changelog.lisp [hunk 6]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+     (:li "Add " (:nxref :command 'nyxt:toggle-prompt-buffer-focus) ".")
+     (:li "Add " (:nxref :command 'nyxt/mode/prompt-buffer:first-suggestion-within-source) ".")
+     (:li "Add " (:nxref :command 'nyxt/mode/prompt-buffer:last-suggestion-within-source) ".")))
+
+  (:nsection
+    :title "New modes"
+    (:ul
+     (:li "New " (:nxref :mode 'nyxt/mode/record-input-field:record-input-field-mode)
+          " to record and restore input fields.")
+     (:li (:nxref :command 'nyxt/mode/repl:repl "lisp REPL")
```
Comment (Ambrevar) [inline source/changelog.lisp:440]: "Lisp"

File: source/changelog.lisp [hunk 7]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+          " is fully re-thought into a multiple-pane environment with debugging, value
```
Comment (Ambrevar) [inline source/changelog.lisp:441]: "overhauled" or "redesigned" instead of "re-thought"

File: source/changelog.lisp [hunk 8]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+          " is incorporated into Nyxt core, with its settings residing in "
+          (:nxref :class-name 'modable-buffer) ".")
+     (:ul
+      (:li "The new "
+           (:nxref :slot 'apply-all-matching-auto-rules-p :class-name 'modable-buffer)
+           " allows you to apply all the matching auto-rules instead of the most
+specific one.")
+      (:li "There are default rules for Gopher, Gemini, and Nyxt-internal-pages.")
+      (:li "The rules file is now moved to auto-rules.lisp (instead of the old
+auto-mode-rules.lisp)."))
```
Comment (Ambrevar) [inline source/changelog.lisp:477]: Use `(:code ...)` or the like for file names.

File: source/changelog.lisp [hunk 9]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+     (:li "The buffer hierarchy has been redesigned.  Now " (:nxref :class-name 'buffer)
+          " is a minimal class and instantiating such a buffer is only useful if you need
+a dummy buffer. " (:nxref :class-name 'web-buffer)
+          " inherits from a mix of specialized buffer subclasses, such as "
+          (:nxref :class-name 'modable-buffer) " and " (:nxref :class-name 'input-buffer)
+          ".  For the full list, see the "
+          (:nxref :class-name 'buffer) " class documentation and browse its subclasses.")
+     (:li (:code "application-mode") " is now "
+          (:nxref :mode 'nyxt/mode/passthrough:passthrough-mode) ".")
+     (:li (:code "web-mode") " is no more.  Instead much of its features has been  moved to the new "
```
Comment (Ambrevar) [inline source/changelog.lisp:496]: "have been move" (also spurious space)

File: source/changelog.lisp [hunk 10]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+          (:nxref :class-name 'nyxt/mode/hint:hint-mode :slot 'style) ".")
+     (:li "Deprecate " (:code "nyxt/mode/hint:highlighted-box-style") " and merge into "
+          (:nxref :class-name 'nyxt/mode/hint:hint-mode :slot 'style) ".")
+     (:li "New " (:nxref :function 'describe-mode) " command (an alias for "
+          (:nxref :function 'describe-class) ").")
+     (:li (:nxref :function 'describe-command) " became an alias for "
+          (:nxref :function 'describe-function) ".")
+     (:li (:code "nyxt/mode/prompt-buffer:return-selection") " renamed to "
+          (:nxref :command 'nyxt/mode/prompt-buffer:run-action-on-return)
+          ".  The default keybinding is the same.")
```
Comment (Ambrevar) [inline source/changelog.lisp:519]: Is it useful to mention that the default keybinding is the same?  I would only mention something when it has changed.

**Comment by Ambrevar** [comment]
Wait, I'm reviewing.

File: source/changelog.lisp [hunk 11]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+also added.")
+     (:li "New " (:nxref :function 'prompt1) " helper.")
+     (:li "New " (:nxref :package :theme) " library.")
+     (:li "Input processing is now easier to customize with "
+          (:nxref :slot 'command-dispatcher :class-name 'window)
+          " and " (:nxref :slot 'input-skip-dispatcher :class-name 'window) ".")
+     (:li "Rename buffer slot " (:code "load-status") " to "
+          (:nxref :slot 'status :class-name 'buffer) ".")
+     (:li "The core " (:code "nyxt") " packages are now locked"
+          "to prevent against accidental clobbering from the user side.")
```
Comment (Ambrevar) [inline source/changelog.lisp:625]: Missing space between the two lines?

File: source/changelog.lisp [hunk 12]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+     (:li "General purpose helpers can be found in the "
+          (:nxref :package :nyxt/utilities) " package.")
+     (:li "New " (:code "nxref") " Spinneret tag for cross-referencing.")
+     (:li (:nxref :macro 'if-confirm)
+          " now allows configuring its yes/no options and can return booleans.")
+     (:li "Move download hooks to " (:nxref :class-name 'nyxt/mode/download:download)
+          " enabling proper typing and adding handlers to them.")
+     (:li "Spinneret tags like " (:code ":nxref") ", " (:code ":nbutton") ", " (:code ":ninput") ", "
+          (:code ":nselect") ", " (:code ":ncode") ", " (:code ":nsection") ", " (:code ":ntoc")
+          " for better internal pages markup enhansed by Lisp image data.")
```
Comment (Ambrevar) [inline source/changelog.lisp:641]: typo: "enhanced"

Comment (Ambrevar) [reply source/changelog.lisp:641]: What does "Lisp image data" mean?

File: source/changelog.lisp [hunk 13]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+          " to be an arbitrary function.")
+     (:li "New package nicknames:"
+          (:ul
+           (:li  (:code "time") " for " (:code "local-time"))
+           (:li  (:code "types") " for " (:code "trivial-types"))
+           (:li  (:code "sym") " for " (:code "nsymbols"))))
+     (:li "The third value in the " (:nxref :function 'prompter:object-attributes)
+          " attribute list is interpreted as display HTML for the suggestion. See the
+color-picker support as an example application for this feature.")
+     (:li "New " (:nxref :function 'match-port) " URL designator predicate for auto-mode rules.")))
```
Comment (Ambrevar) [inline source/changelog.lisp:662]: We don't have `auto-mode` anymore :)

File: source/changelog.lisp [hunk 14]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+
+  (:nsection
+    :title "Bug fixes"
+    (:ul
+     (:li "Lisp run with the --script or --eval command line arguments"
```
Comment (Ambrevar) [inline source/changelog.lisp:667]: "Lisp code"

File: source/changelog.lisp [hunk 15]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+          " now default to the " (:code "nyxt-user") " package.")
```
Comment (Ambrevar) [inline source/changelog.lisp:668]: "defaults"

File: source/changelog.lisp [hunk 16]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+          " by gracefully handling pages that are not yet done loading.")
+     (:li "Fix full-screening event handling — status buffer no longer goes off-sync with
+the full-screened page/video.")
+     (:li "Startup is more robust against corrupted history files.")
+     (:li "VI insert mode is triggered in more cases where it should be triggered.")
+     (:li "Invoke the right WebKit command when cutting text with "
+          (:nxref :function 'ffi-buffer-cut) ".")
+     (:li "Fix the display of history suggestions when going forward in history.")
+     (:li "Security: all the non-ASCII domain names are shown as IDN punycodes in addition
+to aesthetic display in status buffer.")
```
Comment (Ambrevar) [inline source/changelog.lisp:687]: "in the status buffer".

File: source/changelog.lisp [hunk 17]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+     (:li "The canceled page requests are stored to history, making it more consistent.")
+     (:li "Prompt buffer updated for intuitive matching with new algorithms and
+settings. It also relies on the external " (:nxref :package :prompter) " library.")
+     (:li "Add Flatpak build.")
```
Comment (Ambrevar) [inline source/changelog.lisp:691]: Wrong section! :)

File: source/changelog.lisp [hunk 18]
```diff
@@ -327,462 +330,368 @@ SLY install.")
...
+     (:li "Startup is more robust against corrupted history files.")
+     (:li "VI insert mode is triggered in more cases where it should be triggered.")
+     (:li "Invoke the right WebKit command when cutting text with "
+          (:nxref :function 'ffi-buffer-cut) ".")
+     (:li "Fix the display of history suggestions when going forward in history.")
+     (:li "Security: all the non-ASCII domain names are shown as IDN punycodes in addition
+to aesthetic display in status buffer.")
+     (:li "The canceled page requests are stored to history, making it more consistent.")
+     (:li "Prompt buffer updated for intuitive matching with new algorithms and
+settings. It also relies on the external " (:nxref :package :prompter) " library.")
```
Comment (Ambrevar) [inline source/changelog.lisp:690]: Wrong section?  Also seems to repeat what was said before.

File: source/changelog.lisp (hunk 5 above)
Comment (aartaka) [reply source/changelog.lisp:420]: But wait, it's it the rule that the last punctuation counts as sentence-ending, if it ends the parenthesided part?

File: source/changelog.lisp (hunk 12 above)
Comment (aartaka) [reply source/changelog.lisp:641]: I'll rephrase it to "compiler data", is that clearer?

Comment (Ambrevar) [reply source/changelog.lisp]: Too arcane for the casual reader.  Maybe "with semantic highlighting"?

File: source/changelog.lisp (hunk 5 above)
Comment (Ambrevar) [reply source/changelog.lisp:424]: Here you'd write

```
... results. (Thanks fo @BlueFlo0d!)
```

The capital in "Thanks" indicates that this is a separate sentence.  So there must be a `.` before that.

File: source/changelog.lisp (hunk 12 above)
Comment (aartaka) [reply source/changelog.lisp]: But wait, that's a "Programming interface" section, it's intended for an advanced reader...

File: source/changelog.lisp (hunk 5 above)
Comment (aartaka) [reply source/changelog.lisp:424]: I'll decapitalize (liberte, egalite, fraternite :P) it then!

**Comment by aartaka** [comment]
Merging?

File: source/changelog.lisp (hunk 12 above)
Comment (Ambrevar) [reply source/changelog.lisp]: Still, "compiler data" is too vague.  Maybe "semantic from the compiler metadata"?

**Comment by Ambrevar** [comment]
Looks good to me!
